### DISCUSSION TITLE
How to get layout of managed object conveniently?

### DISCUSSION THREAD

**Discussion by huoyaoyuan** [dotnet/runtime#94481] [Q&A] [answered]
During development of vm code, it's often necessary to confirm object layout. Currently I'm relying on checks in corelib.h, which is inconvenient and can miss other architectures.
Is there any tooling to get layout of each field, or is there any doc to explain the layout rule?
VS17.8 is adding object size and alignment hint in IntelliSense for C++.

@jkotas any advise?

**Comment by jkotas** [comment]
We have been trying to minimize the number of mirrored structures and keep the data structures that need to be mirrored simple. corelib.h is about 2x smaller compared to what it was in .NET Core 1.0 days. 

There is no silver bullet for computing the matching layouts for complex data structures. Both the managed and unmanaged field layout algorithms are architecture specific.

**Comment by jkotas** [comment]
> there any doc to explain the layout rule?

The sequential field layout is simple. The auto layout - that everything that contains GC references uses - is reordering the fields to maximize packing (primitive fields of the same size are reordered together) and improve GC reference locality (all GC references are reordered together).

**Reply by DaZombieKiller** [reply]
> all GC references are reordered together

Is there any particular order that the algorithm currently stores GC references in after they have been grouped? Do they retain the declaration order or something else?

**Reply by jkotas** [reply]
They retain the declaration order.

**Reply by huoyaoyuan** [reply]
I'm confused with the reorder of unmanaged fields with different sizes right now. Namely, an object with `object, object, IntPtr, IntPtr, int, bool, bool`.

**Reply by MichalPetryka** [reply, marked as answer]
> Namely, an object with `object, object, IntPtr, IntPtr, int, bool, bool`.

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoagYjxwBtGBDYRmAAgBMBLAZzYdq1AJJ5+ABxhgMAOgASMFpIAUeGAHdOAFRj8MqgJRGA3CKpEAzJ1K79GagG9qnN7ZsRgAKxkZOLJwAvJwamsbmVO4enF6+spzAwaFaEa7u1pziGAAKGFCcYMlkkdGZ2XkF3MkkpRk2vHj+XCFWdW6ZwBAQjJwAZsl9LIz8MO0xXT2cAObJ+Thj1AC+QA===

**Reply by huoyaoyuan** [reply]
@MichalPetryka good solution here.
There is also `dumpobj` command for WinDbg, just no convenient way within VS. Anyway, using diagnostic tools to get the layout should be expected way.

**Comment by The-Futurist** [comment]
@huoyaoyuan 
> During development of vm code, it's often necessary to confirm object layout. Currently I'm relying on checks in corelib.h, which is inconvenient and can miss other architectures. Is there any tooling to get layout of each field, or is there any doc to explain the layout rule? VS17.8 is adding object size and alignment hint in IntelliSense for C++.
> 
> @jkotas any advise?

You can create custom IL to enumerate all fields and then get their [offsets](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.ldflda?view=net-8.0) and sizes, that's something done here for example:  

https://github.com/dotnet/runtime/discussions/100487#discussioncomment-8976548

If you copy that code and play around with it, you'll see it works OK for `ref` or `value` types (the code isn't as good as it could be, no pinning and so on but you get the idea).

**Reply by huoyaoyuan** [reply]
Well this question is specific for CoreLib. If things are getting wrong, the CLR just refuses to execute any managed code.
I can just copy the data structure and execute in another project instead.
